corehq/apps/userreports/pillow.py

```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
             # remove it until the next bootstrap call
             self.table_adapters_by_domain[domain].remove(table)
 
+    def process_changes_chunk(self, pillow_instance, changes):
+        failed_changes = set()
+
+        def get_docs(_changes, domain):
+            # get docs in bulk and validate revision
+            if not _changes:
+                return []
+            # break up doctype
+            _changes_by_doc_store = defaultdict(list)
+            for change in _changes:
+                _changes_by_doc_store[change.document_store].append(change)
+
+            # query
+            docs = []
+            for doc_store, d_changes in six.iteritems(_changes_by_doc_store):
+                docs.extend(list(doc_store.iter_documents([change.id for change in d_changes])))
+
+            # catch missing docs
+            docs_by_id = {doc['_id']: doc for doc in docs}
+            for change in _changes:
+                if change.id not in docs_by_id:
+                    # we need to capture DocumentMissingError which is not possible in bulk
+                    #   so let pillow fall back to serial mode to capture the error for missing docs
+                    failed_changes.add(change)
+                    continue
+                try:
+                    ensure_matched_revisions(change, docs_by_id.get(change.id))
+                except DocumentMismatchError:
+                    failed_changes.add(change)
+            good_changes = set(_changes) - failed_changes
+            return [docs_by_id.get(change.id) for change in good_changes]
+
+        self.bootstrap_if_needed()
+
+        # break up changes by domain
+        changes_by_domain = defaultdict(list)
+        changes_by_id = {change.id: change for change in changes}
+        for change in changes:
+            if change.metadata.domain and change.metadata.domain in self.table_adapters_by_domain:
+                # if no domain we won't save to any UCR table
+                changes_by_domain[change.metadata.domain].append(change)
+
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            rows_to_save_by_adapter = defaultdict(list)
+            adapters = list(self.table_adapters_by_domain[domain])
+            to_delete = {change for change in changes_chunk if change.deleted}
```

Comment:
I think that to_delete only ever needs the ID. so this could be changed to `{change.id for change in changes_chunk if change.deleted}` which would remove some list comprehensions later in the code